src/libpmem2/badblocks_ndctl.c

```diff
@@ -806,3 +807,220 @@ badblocks_clear_all(const char *file)
 
 	return ret;
 }
+
+int
+pmem2_badblock_iterator_new(const struct pmem2_source *src,
+		struct pmem2_badblock_iterator **pbb)
+{
+	ASSERTne(pbb, NULL);
+
+	struct ndctl_ctx *ctx;
+	struct ndctl_region *region;
+	struct ndctl_namespace *ndns;
+	os_stat_t st;
+	int ret = -1;
+
+	if (ndctl_new(&ctx)) {
+		ERR("!ndctl_new");
+		return PMEM2_E_ERRNO;
+	}
+
+	if (os_fstat(src->fd, &st)) {
+		ERR("!fstat %i", src->fd);
+		ret = PMEM2_E_ERRNO;
+		goto error_ndctl_unref;
+	}
+
+	int rv = ndctl_region_namespace(ctx, &st, &region, &ndns);
+	if (rv) {
+		LOG(1, "getting region and namespace failed");
+		ret = rv;
+		goto error_ndctl_unref;
+	}
+
+	if (ndns == NULL) {
+		ERR("cannot get namespace");
+		goto error_ndctl_unref;
+	}
+
+	/*
+	 * Only the new NDCTL versions have the namespace badblock iterator.
+	 * When compiled with older versions, the library needs to rely
+	 * on the old region interface.
+	 */
+	if (ndctl_namespace_get_mode(ndns) == NDCTL_NS_MODE_FSDAX) {
+		struct pmem2_badblock_iterator *bb =
+			Zalloc(sizeof(struct pmem2_badblock_iterator));
+		if (bb == NULL) {
+			ERR("!Zalloc");
+			ret = PMEM2_E_ERRNO;
+			goto error_ndctl_unref;
+		}
+
+		bb->ndctl_has_namespace_badblock_iterator = 1;
+		bb->ndns = ndns;
+		*pbb = bb;
+
+	} else {
+		if (region == NULL) {
+			ERR("cannot get region");
+			goto error_ndctl_unref;
+		}
+
+		unsigned long long ns_beg, ns_size, ns_end;
+
+		if (badblocks_get_namespace_bounds(region, ndns,
+		    &ns_beg, &ns_size)) {
+			LOG(1, "cannot read namespace's bounds");
+			goto error_ndctl_unref;
+		}
+
+		ns_end = ns_beg + ns_size - 1;
+
+		LOG(10,
+			"namespace: begin %llu, end %llu size %llu (in 512B sectors)",
+			B2SEC(ns_beg), B2SEC(ns_end + 1) - 1, B2SEC(ns_size));
+
+		struct pmem2_badblock_iterator *bb =
+			Zalloc(sizeof(struct pmem2_badblock_iterator));
+		if (bb == NULL) {
+			ERR("!Zalloc");
+			ret = PMEM2_E_ERRNO;
+			goto error_ndctl_unref;
+		}
+
+		bb->ndctl_has_namespace_badblock_iterator = 0;
+		bb->region = region;
+		bb->ns_beg = ns_beg;
+		bb->ns_end = ns_end;
+		*pbb = bb;
+	}
+
+	ret = 0;
+
+error_ndctl_unref:
+	ndctl_unref(ctx);
+
+	return ret;
+}
+
+void
+pmem2_badblock_iterator_delete(struct pmem2_badblock_iterator **pbb)
+{
+	ASSERTne(pbb, NULL);
+	Free(*pbb);
+	*pbb = NULL;
+}
+
+/*
+ * pmem2_badblock_next_namespace -- (internal) version of pmem2_badblock_next()
+ *                                  called for ndctl with namespace badblock
+ *                                  iterator
+ */
+static int
+pmem2_badblock_next_namespace(struct pmem2_badblock_iterator *pbb,
+				struct pmem2_badblock *bb)
+{
+	LOG(3, "pbb %p bb %p", pbb, bb);
+
+	ASSERTne(pbb, NULL);
+	ASSERTne(bb, NULL);
+
+	struct badblock *bbn;
+
+	if (pbb->n_iter++ == 0) {
```

Comment:
While highly improbable I think it's technically possible to overflow this variable with 4TB device/file.

Reply:
Done